substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/util/Utf8.java

```diff
@@ -226,12 +225,20 @@ public static final class WrappedAsciiCString implements CharSequence {
         private final int length;
         private final int hashCode;
 
-        public WrappedAsciiCString(CCharPointer chars, int length, int hashCode) {
+        private WrappedAsciiCString(CCharPointer chars, int length, int hashCode) {
             this.chars = chars;
             this.length = length;
             this.hashCode = hashCode;
+        }
+
+        private static WrappedAsciiCString create(CCharPointer chars, int length, int hashCode) {
+            final String s = utf8ToString(chars);
+            if (s == null || s.length() != length || s.hashCode() != hashCode) {
+                return null;
+            }
```

Comment:
This always transforms the characters to a `java.lang.String` instead of doing so only as an assertion, losing the advantage of `WrappedAsciiCString` entirely. What does this accomplish anyway? If the string contains non-ASCII characters, the caller will instead invoke `utf8ToString` and add any additional checks there.